### PR TITLE
Allow disabling DropDownButton with optional hint for disabled state

### PR DESCRIPTION
**Pull request by escamoteur**
A quite common pattern in App that allows to filter Data, is selecting properties in multiple DropDownButtons where the available properties for the next DropDownbutton are only avaliable and dependending  on the value selected with the upper DropDownButton.
Currently it was not possible to disable a DropDownButton if you had no data ready for its items.. Therefore I make this PR.

I tried if it is possible to pass null or an empty List for `items` but both threw exceptions and did not disabling tapping on the button.

Now if you pass null for `items`it will disable the button and draw the down arrow in a lighter color.
 
Additionally you can provide an optional `hintDisabled' widget which will be displayed instead of `hint` if the button is disabled like here:

image

by updating `items` with a valid list it automatically get:

image

If no `hintDisabled` is provided `hint`will be displayed in disabled state too.

### CODE REVIEW

**Comment by Hixie** [comment]
@escamoteur Looks like there was some issue where a formatter got a bit excited on this code; any chance you can undo the formatting changes to make the patch easier to review? Thanks.

File: packages/flutter/lib/src/material/dropdown.dart [hunk 1]
```diff
@@ -464,6 +464,7 @@ class DropdownButton<T> extends StatefulWidget {
   /// Creates a dropdown button.
   ///
   /// The [items] must have distinct values and if [value] isn't null it must be among them.
+  /// If [items] is null the button is disabled and shows the optional [disabledHint] and the down arrow is drawn as disabled
```
Comment (Hixie) [inline packages/flutter/lib/src/material/dropdown.dart]: nit: please wrap comments to about 80 characters and finish sentences with a period.

Comment (Hixie) [reply packages/flutter/lib/src/material/dropdown.dart]: Consider something like:
```
  /// If [items] is null, the button will be disabled, the down arrow will be grayed out, and
  /// the [disabledHint] will be shown (if provided).
```

File: packages/flutter/lib/src/material/dropdown.dart [hunk 2]
```diff
@@ -472,14 +473,16 @@ class DropdownButton<T> extends StatefulWidget {
...
     this.elevation = 8,
     this.style,
     this.iconSize = 24.0,
     this.isDense = false,
-  }) : assert(items != null),
-       assert(value == null || items.where((DropdownMenuItem<T> item) => item.value == value).length == 1),
-      super(key: key);
+  })  : assert(items == null
+      || value == null
+      || items.where((DropdownMenuItem<T> item) => item.value == value).length == 1),
```
Comment (Hixie) [inline packages/flutter/lib/src/material/dropdown.dart]: the formatting on this assert is a bit confusing. Consider something like:

```dart
  })  : assert(items == null || value == null ||
               items.where((DropdownMenuItem<T> item) => item.value == value).length == 1),
```

**Comment by Hixie** [comment]
I think it would be more idiomatic if the dropdown button was disabled when `onChanged` was null, rather than `items`.

**Comment by Hixie** [comment]
This will need tests.

File: packages/flutter/lib/src/material/dropdown.dart (hunk 1 above)
Comment (escamoteur) [reply packages/flutter/lib/src/material/dropdown.dart]: done

File: packages/flutter/lib/src/material/dropdown.dart (hunk 2 above)
Comment (escamoteur) [reply packages/flutter/lib/src/material/dropdown.dart]: done

**Comment by escamoteur** [comment]
Fixed the formatting as requested. I also added that if `onChanged == null` it disabled the control too. I think letting `items==null` make it disabled makes a lot of sense if you get this value updated by a `setState` quite often.

File: packages/flutter/lib/src/material/dropdown.dart [hunk 3]
```diff
@@ -34,14 +34,14 @@ class _DropdownMenuPainter extends CustomPainter {
...
-       ).createBoxPainter(),
-       super(repaint: resize);
+    // If you add an image here, you must provide a real
+    // configuration in the paint() function and you must provide some sort
+    // of onChanged callback here.
+      color: color,
+      borderRadius: new BorderRadius.circular(2.0),
+      boxShadow: kElevationToShadow[elevation]
+  ).createBoxPainter(),
+        super(repaint: resize);
```
Comment (Hixie) [inline packages/flutter/lib/src/material/dropdown.dart]: looks like you still have some indentation changes here

File: packages/flutter/lib/src/material/dropdown.dart [hunk 4]
```diff
@@ -393,7 +393,7 @@ class DropdownMenuItem<T> extends StatelessWidget {
     this.value,
     @required this.child,
   }) : assert(child != null),
-       super(key: key);
+        super(key: key);
```
Comment (Hixie) [inline packages/flutter/lib/src/material/dropdown.dart]: and some here and below

File: packages/flutter/lib/src/material/dropdown.dart [hunk 5]
```diff
@@ -464,22 +464,25 @@ class DropdownButton<T> extends StatefulWidget {
...
+    this.disabledHint,
     @required this.onChanged,
     this.elevation = 8,
     this.style,
     this.iconSize = 24.0,
     this.isDense = false,
-  }) : assert(items != null),
-       assert(value == null || items.where((DropdownMenuItem<T> item) => item.value == value).length == 1),
-      super(key: key);
+  })  : assert(items == null || value == null   ||
```
Comment (Hixie) [inline packages/flutter/lib/src/material/dropdown.dart]: nit: extraneous spaces before the final `||`

File: packages/flutter/lib/src/material/dropdown.dart [hunk 6]
```diff
@@ -492,6 +495,9 @@ class DropdownButton<T> extends StatefulWidget {
   /// Displayed if [value] is null.
   final Widget hint;
 
+  /// Displayed if [items] is null.
```
Comment (Hixie) [inline packages/flutter/lib/src/material/dropdown.dart]: update documentation since it also displays if onChanged is null.

File: packages/flutter/lib/src/material/dropdown.dart [hunk 7]
```diff
@@ -563,13 +569,17 @@ class _DropdownButtonState<T> extends State<DropdownButton<T>> with WidgetsBindi
...
-    assert(widget.value == null ||
-      widget.items.where((DropdownMenuItem<T> item) => item.value == widget.value).length == 1);
-    _selectedIndex = null;
-    for (int itemIndex = 0; itemIndex < widget.items.length; itemIndex++) {
-      if (widget.items[itemIndex].value == widget.value) {
-        _selectedIndex = itemIndex;
-        return;
+    if (widget.items != null) {
+      assert(widget.value == null ||
+          widget.items
```
Comment (Hixie) [inline packages/flutter/lib/src/material/dropdown.dart]: nit: indent this a bit more so it's vertically aligned (yes, the original code was bad for this too, oops)

File: packages/flutter/lib/src/material/dropdown.dart [hunk 8]
```diff
@@ -613,27 +623,31 @@ class _DropdownButtonState<T> extends State<DropdownButton<T>> with WidgetsBindi
     return math.max(_textStyle.fontSize, math.max(widget.iconSize, _kDenseButtonHeight));
   }
 
+  // just so we don't have to repeat this test and make it easier to understand
+  bool get _enabled => widget.items != null && onChanged != null;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/dropdown.dart]: no need for the comment here

File: packages/flutter/lib/src/material/dropdown.dart [hunk 9]
```diff
@@ -613,27 +623,31 @@ class _DropdownButtonState<T> extends State<DropdownButton<T>> with WidgetsBindi
...
+
   @override
   Widget build(BuildContext context) {
     assert(debugCheckHasMaterial(context));
 
     // The width of the button and the menu are defined by the widest
     // item and the width of the hint.
-    final List<Widget> items = new List<Widget>.from(widget.items);
+    final List<Widget> items =
+    _enabled ? new List<Widget>.from(widget.items) : <Widget>[];
```
Comment (Hixie) [inline packages/flutter/lib/src/material/dropdown.dart]: nit: indent by 4 if you wrap after a =

File: packages/flutter/lib/src/material/dropdown.dart [hunk 10]
```diff
@@ -644,17 +658,23 @@ class _DropdownButtonState<T> extends State<DropdownButton<T>> with WidgetsBindi
...
               size: widget.iconSize,
               // These colors are not defined in the Material Design spec.
-              color: Theme.of(context).brightness == Brightness.light ? Colors.grey.shade700 : Colors.white70
+              color: _enabled
+                  ? Theme.of(context).brightness == Brightness.light
+                  ? Colors.grey.shade700
+                  : Colors.white70
+                  : Theme.of(context).brightness == Brightness.light
+                  ? Colors.grey.shade400
+                  : Colors.white10,
```
Comment (Hixie) [inline packages/flutter/lib/src/material/dropdown.dart]: consider factoring this out into a separate method that uses "if" statements, it would be easier to read

File: packages/flutter/lib/src/material/dropdown.dart [hunk 11]
```diff
@@ -673,7 +693,7 @@ class _DropdownButtonState<T> extends State<DropdownButton<T>> with WidgetsBindi
             child: new Container(
               height: 1.0,
               decoration: const BoxDecoration(
-                border: const Border(bottom: const BorderSide(color: const Color(0xFFBDBDBD), width: 0.0))
+                  border: const Border(bottom: const BorderSide(color: const Color(0xFFBDBDBD), width: 0.0))
```
Comment (Hixie) [inline packages/flutter/lib/src/material/dropdown.dart]: nit: unnecessary whitespace change

**Comment by Hixie** [comment]
This will need tests.

**Comment by Hixie** [comment]
(To write tests, look in the `//packages/flutter/test/material/` directory for the existing tests for this widget, and follow the same pattern.)

File: packages/flutter/lib/src/material/dropdown.dart (hunk 10 above)
Comment (escamoteur) [reply packages/flutter/lib/src/material/dropdown.dart]: Done

File: packages/flutter/lib/src/material/dropdown.dart (hunk 9 above)
Comment (escamoteur) [reply packages/flutter/lib/src/material/dropdown.dart]: done

File: packages/flutter/lib/src/material/dropdown.dart (hunk 11 above)
Comment (escamoteur) [reply packages/flutter/lib/src/material/dropdown.dart]: done

**Comment by escamoteur** [comment]
I tried to add a test. unfortunately I could not run the test because on this branch other tests of this control are broken,

**Comment by Hixie** [comment]
@escamoteur Will you be able to continue working on this? It's totally fine if not; we can take it over (though it might take longer). I'm just asking because I'm trying to clean out our review queue.

**Comment by escamoteur** [comment]
If you could take over that would be great, but I would be very disappointed if this would get lost after the work I done on it

**Comment by Hixie** [comment]
Definitely won't get lost, we use the "needs love" label to look at closed PRs that we should be taking further. It just might take a long time, we don't have much time to give to these PRs yet. :-(

**Comment by escamoteur** [comment]
@Hixie for now leave it open, I will try to get back to it in October

**Comment by mklim** [comment]
I've adopted this into flutter/flutter#22539. Thanks for your contribution @escamoteur!

**Comment by mklim** [comment]
Moved back from flutter/flutter#22539 to here, incorporated the early feedback from that PR.

**Comment by Hixie** [comment]
LGTM modulo minor nits

**Comment by mklim** [comment]
@escamoteur, do you mind if I force push a [squashed version of these changes](https://github.com/mklim/flutter/commit/2a7a7674538711104be9252e6873de8e8cc5b8f7) to this branch? The older commits and multiple merges are causing issues with how `flutter version` is picked up. Combining the commits and rebasing onto master fixes it, but will overwrite the existing history.

**Comment by escamoteur** [comment]
Sure go on
